raft/raft.go

```diff
@@ -796,7 +869,7 @@ func (r *raft) becomeLeader() {
 	// uncommitted log quota. This is because we want to preserve the
 	// behavior of allowing one entry larger than quota if the current
 	// usage is zero.
-	r.reduceUncommittedSize([]pb.Entry{emptyEnt})
+	r.uncommittedSize = 0
```

Comment:
There's something fishy here. `getUncommittedSize` returns `0` when `r.uncommittedSize == 0`. So aren't we accidentally disabling the tracking here, since we won't ever add to `r.uncommittedSize` again? If so, let's make sure this is also exercised in a test.

Even if the above problem isn't actually occurring for some reason I don't understand, isn't resetting here open up the possibility of each leader amassing (say) 10mb of uncommitted log, losing leadership to a second node which adds another 10mb of uncommitted log, loses leadership back, adds another 10mb, etc, for essentially unbounded log growth? This would need a replication factor of 5 (if the log is to be "truly" uncomitted) but still.

Reply:
Yes, the gap in tracking that you proposed is real. It's been since this tracking was introduced because we didn't want to scan the uncommitted log on leadership changes.

I've cleaned some of this logic up so that the code is more clear, but that didn't resolve the problem. Let's extract that discussion from this PR, since it's not a new concern.